ophyd/status.py

```diff
@@ -612,19 +737,13 @@ def wait(status, timeout=None, *, poll_rate=0.05):
     RuntimeError
         If the status failed to complete successfully
     """
-    t0 = time.time()
-
-    def time_exceeded():
-        return timeout is not None and (time.time() - t0) > timeout
-
-    while not status.done and not time_exceeded():
-        time.sleep(poll_rate)
-
-    if status.done:
-        if status.success is not None and not status.success:
-            raise RuntimeError('Operation completed but reported an error: {}'
-                               ''.format(status))
-    elif time_exceeded():
-        elapsed = time.time() - t0
-        raise TimeoutError('Operation failed to complete within {} seconds '
-                           '(elapsed {:.2f} sec)'.format(timeout, elapsed))
+    # It would probably be more useful to just return status.wait(timeout)
+    # directly rather than chaining a RuntimeError on the end here. Is it worth
+    # maintaining back-compat on this?
```

Comment:
Shouldn't we then fix the built-in EpicsMotor and such to `set_exception`?

Reply:
Yes. I think we can do this incrementally in separate PRs. I expect it will take some thought and significant effort to decide on the right exceptions and error messages in each case. As it stands, anything that uses the old path (`status._finished(success=False)`) will cause an `UnknownStatusFailure` exception to be raised.

The question here is whether it's better to let `UnknownStatusFailure` (and, in the future, more specific and useful exceptions) through here or to chain them behind a general `RuntimeError` in the name of backward-compatibility. We could make `UnknownStatusFailure` a subclass of `RuntimeError` if we like. Either way, I think it's probably useful to let the original exception through, removing

```py
    except WaitTimeoutError as exc:
        raise TimeoutError(*exc.args)
    except Exception as exc:
        raise RuntimeError('Operation completed but reported an error: {}'
                           ''.format(status)) from exc
```
especially since readers of these tracebacks are often not like to read up through a chained exception.

Reply:
Makes sense. I'd be in favor of the `UnknownStatusFailure(RuntimeError)` (subclass) option.

Reply:
I tried that, but it turns out the tests still fail with that approach, because `TimeoutError` is caught and retained as `RuntimeError` on master but would just be raised directly with the new approach.

I think it is more useful to raise the exception set by `set_exception` than to chain or wrap it in `RuntimeError`. The error from `wait(status)` should be consistent with the error from `status.wait()`. I now think this is worth an API change. I would guess that raising a different exception from the `wait` would not have much impact on user code. I’m not sure it is used at all.

Reply:
I suppose I didn't quite understand the intricacies of it, but I definitely trust your intuition.